File: src/main/java/uk/gov/justice/digital/hmpps/riskprofiler/services/DefaultFileService.java
package uk.gov.justice.digital.hmpps.riskprofiler.services;

import com.amazonaws.util.IOUtils;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.stereotype.Component;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Arrays;
import java.util.Comparator;

@Component
@Slf4j
@ConditionalOnProperty(name = "file.process.type", havingValue = "file")
public class DefaultFileService implements FileService {

    public PendingFile getLatestFile(String path) {
        File folder = new File(path);
        File[] listOfFiles = folder.listFiles();

        if (listOfFiles != null) {
            log.info("Found {} files in {}", listOfFiles.length, path);
            return Arrays.stream(listOfFiles)
                    .filter(f -> !f.isDirectory())
                    .max(Comparator.comparing(File::lastModified))
                    .map( f -> {
                        try {
                            return PendingFile.builder()
                                    .fileName(f.getName())
                                    .fileTimestamp(LocalDateTime.ofInstant(Instant.ofEpochMilli(f.lastModified()), ZoneId.systemDefault()))
                                    .data(IOUtils.toByteArray(new FileInputStream(f)))
                                    .build();
                        } catch (IOException e) {
                            return null;
                        }
                    })
                    .orElse(null);

        }
        log.info("Found {} files in {}", 0, path);
        return null;
    }

    @Override
    public void deleteHistoricalFiles(String fileLocation) {
        File folder = new File(fileLocation);
        File[] listOfFiles = folder.listFiles();

        if (listOfFiles != null) {
            log.info("Housekeeping- found {} files in {}", listOfFiles.length, fileLocation);
            Arrays.stream(listOfFiles).sorted(Comparator.comparing(File::lastModified).reversed()).skip(2).forEach(file -> {
                file.delete();
                log.info("Deleted file {} ", fileLocation + "/" + file.getName());
            });
        }
    }


}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/riskprofiler/services/DefaultFileService.kt
package uk.gov.justice.digital.hmpps.riskprofiler.services

import com.amazonaws.util.IOUtils
import org.slf4j.LoggerFactory
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.stereotype.Component
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.FileType
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.Viper
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.nio.file.Files
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.Arrays
import java.util.stream.Collectors
import kotlin.Comparator

@Component
@ConditionalOnProperty(name = ["file.process.type"], havingValue = "file")
class DefaultFileService : FileService {
  override fun getLatestFile(fileLocation: String, fileType: FileType?): PendingFile? {
    val folder = File(fileLocation)
    val listOfFiles = folder.listFiles()
    if (listOfFiles != null) {
      log.info("Found {} files in {}", listOfFiles.size, fileLocation)
      return Arrays.stream(listOfFiles)
        .filter { f: File -> !f.isDirectory }
        .max(Comparator.comparing { obj: File -> obj.lastModified() })
        .map { f: File ->
          try {
            return@map PendingFile(
              f.name,
              LocalDateTime.ofInstant(Instant.ofEpochMilli(f.lastModified()), ZoneId.systemDefault()),
              if (fileType == FileType.VIPER) getViperFile(f) else IOUtils.toByteArray(FileInputStream(f))
            )
          } catch (e: IOException) {
            return@map null
          }
        }
        .orElse(null)
    }
    log.info("Found {} files in {}", 0, fileLocation)
    return null
  }

  override fun deleteHistoricalFiles(fileLocation: String) {
    val folder = File(fileLocation)
    val listOfFiles = folder.listFiles()
    if (listOfFiles != null) {
      log.info("Housekeeping- found {} files in {}", listOfFiles.size, fileLocation)
      Arrays.stream(listOfFiles).sorted(
        Comparator.comparing { obj: File -> obj.lastModified() }
          .reversed()
      ).skip(2).forEach { file: File ->
        file.delete()
        log.info("Deleted file {} ", fileLocation + "/" + file.name)
      }
    }
  }

  private fun getViperFile(file: File): ByteArray {
    val csv = Files.lines(file.toPath())
      .map { row -> row.split(",") }
      .filter { row -> row[Viper.RECORD_ID] != "0" }
      .map { row -> row.joinToString(",") }
      .collect(Collectors.joining(System.lineSeparator()))

    return csv.toByteArray()
  }

  companion object {
    private val log = LoggerFactory.getLogger(DefaultFileService::class.java)
  }
}
